### PR TITLE
thumbtable : enhance offset after collection changed

### LINKED ISSUE #5735: [darkroom] filmstrip jumps to first pic when changing rating-view
**Describe the bug**

When changing the rating view e.g. between 3* back to 1*, the filmstrip jumps to first pic in collection.

**To Reproduce**
1. Go to 'darkroom' with a collection having approx. 50 pics in 3* and many more in 1*. place your fokus somewhere in the middle of 3* selection
2. Change from 3* down to 1*
3. See error

**Expected behavior**
Same than lighttable, the focus shall stay centred from where the user left one rating to another (if in the new collection is no such pic, stay very near to that position in the collection)

**Platform (please complete the following information):**
* darktable version : 3.1.0+2448~g4b77a5410
* OS                : Linux - kernel 5.7.7-gentoo
* Distro            : Gentoo Base System release 2.6
* Processor         : Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
* Memory            : 32 GB (4 x 8 GB) + 5GB Swap
* Graphics card0    : GeForce GTX 1060 6GB
* Graphics card1    : GeForce RTX 2070 SUPER
* Graphics driver   : nvidia-drivers-440.100-r1
* OpenCL installed  : Yes (opencl-headers-2020.03.13-r1)
* OpenCL activated  : Yes
* Xorg              : xorg-server-1.20.8
* Desktop           : KDE 5
* GTK+              : gtk+-3.24.20
* gcc               : x86_64-pc-linux-gnu-9.3.0
* cflags            : CFLAGS="-march=native-O2-mtune=native-pipe"
* CMAKE_BUILD_TYPE  : "Release"

### LINKED ISSUE #6566: Culling mode with fixed zoom: Film Strip can jump back to start unexpectedly in relation to grouped images
**Update**:

Darktable version 3.4, Windows 10.

1. Load a collection of pictures shot in JPG and RAW.
2. Open culling mode with fixed zoom, with e.g. 2 images.
3. Expand grouped images.
4. Scroll a bit into the collection.
5. Select a JPG (click on it).
6. Collapse grouped images.

--> Filmstrip jumps back to the start and main view gets empty.

It might be argued whether this behavior is as designed, but it kind of breaks my workflow, so I wanted to mention it.
My expectation would be that after step 6 the filmstrip remains at the index of the RAW file (group head) that corresponds to the JPG selected in step 5.

-----------------------------------------------------------------------------------------------------------------------------------------
**Original Bug description for darktable version 3.2.1 with Windows 10.** The steps described in this original post lead to a darktable crash in version 3.4, but this crash is already documented in issue #7445. Keeping this part for the sake of completeness.

**Describe the bug**
In lighttable's culling mode with fixed zoom, if there are grouped images _and_ grouped images are collapsed (button in the top right of the main view part) and then the first ("leftmost") image of the currently shown ones is changed in rating, so that it does not match the current view filter any more, filmstrip "jumps back to the start" and there are no images shown any more until one is selected again in filmstrip. This is frustrating when reviewing a collection of images, because each time after this has happened it is required to first find the image to review next before it is possible to continue. When going through a bigger collection, this can take a lot of time and be quite annoying.

**_Please note_** again that this only happens if grouped images are collapsed - a fact that I only found out recently and after big parts of this error report had already been written. In the following, "image" therefore always refers to "group of images".

For the sake of clarification, the following is an attempt to rephrase the problem in a more formal way:

There is a collection of images currently active as output of the "collect images" module of lighttable, which I'll call **C**. Additionally, there is a view filter active, so that only those images out of **C** that match this filter are shown in the main view as well as in the filmstrip. This subset I'll call **C'**. **C'** is sorted in a certain way, thus it can be seen as a list of images with an index between 1 and **n**, with **n** being the total amount of images in **C'**.
The fixed zoom culling mode is set to show **k** images side by side, so for any selected image at index **i**, images **i** to **(i+k)** are shown.
If now the rating of image at index **i** is changed in a way so that this image is not any more part of **C'** (but only of **C**), then the whole view gets reset and no images are shown any more, as described above. The user needs to scroll through the filmstrip (i.e. through the sorted list that **C'** is) again in order to find the image that is now at **i** to be able to continue where she or he left off before.

**_Note:_**
- If the same is done for one of the other shown images (formally at index **j** with **i** < **j** <= **i+k**), then the problem does not occur - instead the image at **j** is replaced by the next in the list and the view does not get reset.
- The problem does not occur if the change of rating is done on the very first image in **C'** (i = 1).

**To Reproduce**
1. Open lighttable.
2. Load a collection of grouped images (e.g. images shot in JPG + RAW) using the "collect images" module.
3. Important: Make sure "collapse grouped images" is active (button in the top right of the main view).
4. Set a view filter, e.g. "All except rejected" (others can be used to reproduce the behavior if next steps are adapted accordingly)
5. Activate culling mode with fixed zoom and set the number of images to be shown next to each other to 2 (just for simplicitly, same behavior can also be reproduced for bigger number of pictures).
6. Scroll through the images so that the very first image in the list is not shown any more (the error does not seem to occur for that one), then move the mouse over the _left_ of the two shown images and press R. This will remove it from the current view, as it won't match the active filter any more.
**--> The error occurs, filmstrip jumps to the start and the main view gets emptied.**
7. In order to show that this only happens for the left one of the shown images, repeat the procedure, but in step 6 hover the mouse over the _right_ image and press 'R'. Now the error does not occur.

**Expected behavior**
I personally would expect the following behavior:
After removing an image from **C'** by changing its rating, the next image in the list should take its place, i.e. if image at index **i** to **(i+k)** are shown and then **i** gets removed, then the former images **(i+1) to (i+k+1)** should be shown (in fact, these are actually the new images **i to (i+k)** considering **C'**'s new content).

A special case is the one where images **(n-k) to n** are shown and image **(n-k)** gets removed from **C'**.  In this case, the former images **(n-k-1) to (n-1)** (the new images **(n-k) to n** considering **C'**'s new content) should be shown. Around the end of the list and depending on which value **k** has, there may be other corner cases that should be considered in a similar way.

**Screenshots**

**_Good case_**

beforeRejectingRight
Situation before pressing 'R' in order to reject the right image which will remove it from the current **C'**. The mouse cursor was on top of the right image (DSC_4000), but not captured on the screenshot unfortunately.

--------------------------------

afterRejectingRight
Situation immediately after pressing 'R', without any other interaction. This is the good case, behavior as expected. The next image (DSC_4003) has moved to the right slot that was freed by rejecting DSC_4000.

--------------------------------

**_Problematic case_**

beforeRejectingLeft
Same situation as before, but this time the mouse cursor was on top of the left image (DSC_3998).

--------------------------------

afterRejectingLeft
Situation immediately after pressing 'R', without any other interaction. This is the problematic case, filmstrip has "jumped back to the start", main view is empty.

--------------------------------

**Platform:**
 - Darktable Version: 3.2.1
 - OS: Windows 10

**Additional context**

The same matter has been discussed partly already as part of issue  #5735. However, it turned out, that:
1. This particular problem was out of context of that issue.
2. I was lacking some basic understanding of the culling mode and how it works, when I first posted there, which led to an imprecise and uncertain description.

I am now more confident that what I describe is indeed a reproducible behavior and worth its own issue entry. Also, I now found that there is a relation to the grouped images collapsed mode, as prominently mentioned above. This was probably the cause of my own confusion stated in the discussion of #5735 (I was suddenly not able to reproduce the problem any more, which was likely because I had expanded grouped images).

After last time I was (correctly) asked to read the manuals first, I'd like to state that there is of course still a risk that in the end I misunderstood some concept of the culling interface. However, after now reading the darktable 3.0 manual as well as the blog entries about 3.0 and 3.2.1 and spending more time with the culling mode, I still run into this problem regularly. Thus, even in case it _should_ indeed turn out to be my fault in the end, I would at least consider this issue a usability or documentation flaw.

### LINKED ISSUE #7445: v3.4 "R"-key pressed BUG - full crash.
**Describe the bug**

In lighttable mode in full screen preview as Alt+W, i want to reject current foto: i have pressed "r" key and then i have crach window - see screenshot. It occurs every time!

**To Reproduce**

**Expected behavior**

As in older version - NO crash windows and preview next foto on screen.

**Screenshots**
Darktable 3 4 new bug R key

darktable.exe caused an Access Violation at location 00000000637C6382 in module libdarktable.dll Reading from location 0000000000000000.

[darktable_bt_I60GW0.txt](https://github.com/darktable-org/darktable/files/5741507/darktable_bt_I60GW0.txt)

[darktable_bt_U385V0.txt](https://github.com/darktable-org/darktable/files/5741506/darktable_bt_U385V0.txt)

[darktable_bt_X0F8V0.txt](https://github.com/darktable-org/darktable/files/5741504/darktable_bt_X0F8V0.txt)

**Platform (please complete the following information):**
 - Darktable Version: 
 - OS: Windows 7

**Additional context**

### PR DESCRIPTION
**Pull request by AlicVB**
fix #6566 fix #5735 fix #7445
- we avoid as possible to reset to 1 by searching in previous list
- fix a bug in culling if offset is still reset to 1 (no image shown)

### CODE REVIEW

**Comment by AlicVB** [comment]
In fact, this also fix #5735 :)

File: src/libs/tools/global_toolbox.c [hunk 1]
```diff
@@ -18,6 +18,7 @@
 
 #include "common/collection.h"
 #include "common/darktable.h"
+#include "common/debug.h"
```
Comment (TurboGit) [inline src/libs/tools/global_toolbox.c:21]: is this needed ? that's the only change in this file.

Comment (AlicVB) [reply src/libs/tools/global_toolbox.c]: oops, sorry... leftover from a first test... removed !

**Comment by AlicVB** [comment]
and after some discussion, this seems to also fix #7445 :)
Always great to see small changes fixing lot of complex issues ;)

**Comment by TurboGit** [comment]
Perfect! Merging, thanks.

**Comment by TurboGit** [comment]
Merged in master and 3.4.1.

**Pull request merged**